### PR TITLE
Movable insulation logic refactoring

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
The branch improves the logic to calculate movable insulation properties. The ext and int heat transfer coefficients and absorptances only need to calculated once at the beginning of the heat balance surface manager when any movable insulation exists. The subroutines of calculating these properties are transformed from external function calls to loop-wise initializations. The functions are moved to HeatBalanceSurfaceManager and the original movable insulation related files are removed.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

**Comment by xuanluo113** [comment]
@mjwitte This branch is ready for review. CI and performance tests look green. Thanks very much!

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Seems excessive to allocate all of these arrays when typically there will be no surfaces with movable insulation, and at most only a few of the surfaces in the entire model will have it.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: This comment is not relevant here.

File: src/EnergyPlus/HeatBalanceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceManager.cc]: This loop can be skipped if `!AnyMovableInsulation`.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Even if we keep the full arrays for movable insulation, it would be useful to build a list of surfaces with movable insulation and simply loop over that list here instead of looping over every opaque surface.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: This if construct could be eliminated if we had a `SurfAbsExt` array that is intialized to the base construction `AbsorpSolar` and modified or reset as needed in `EvalOutsideMoveableInsulation`.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Similar comment here. `SurfIntAbs` and `SurfIntAbsVis` would eliminate this if and the repeated construction lookups above.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Because exterior movable insulation doesn't have a `Present` variable like inside movable insulation, `SurfMovInsulHExt >0` is being used to test for presence. In the old `EvalOutsideMovableInsulation` function, this was reset to zero when `MovInsulSchedVal <= 0.0` but I don't see that in the new version of this function.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Seems that `EvalInsideMovableInsulation` could manage setting `ITABSF`.  Seems this entire loop could be eliminated then. But what about EMS construction changes.

For your todo list: `ComputeIntThermalAbsorpFactors` and `ComputeIntSWAbsorpFactors` appear to be completely recomputed every timestep? Seems these function are ripe for skipping a zone if nothing has changed with window shade status or movable insulation? Not suggesting you try that here.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Another spot for a list-driven update.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Another place where arrays of surface roughness and absthermal could eliminate this if and repeated lookups from `dataConstruction`.

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 This is a good start. I think there's room for more improvement. See comments below.

**Comment by mjwitte** [comment]
And there are clang formatting check errors - need to apply formatting to all of the changed files.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (amirroth) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: The alternative would be to ... have a single array that indexes to a set of smaller arrays (or even one arrays of structures) sized to fit only the number of surfaces that have `MovableInsulation`? :thinking: How do we feel about that design pattern in general? How broadly should we apply it? I have been thinking about whether we should do something like this for all of the Window specific arrays. :thinking: :thinking:

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (amirroth) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Just factoring this code into a separate loop and guarding it with `if (MovableInsulationAnywhereInTheModel)` will speed up models that have no `MovableInsulation` in them. A "push" implementation like the one @mjwitte suggests will also speed up models that do have `MovableInsulation` in them, although do we really care about such models right now?

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (amirroth) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: We should probably implement inside and outside `MovableInsulation` in similar ways, no? Consistency is the hobgoblin of little minds. And of Amir.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Five HB arrays are added including `SurfAbsSolarExt`, `SurfAbsThermalExt`, `SurfAbsSolarInt`, `SurfAbsThermalInt`, SurfRoughnessExt`. They are initialized with material properties one time at the beginning of `InitSurfaceHeatBalance`, and are set to movable insulation properties (and set back) at run time when `AnyMovableInsulation`.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Five HB arrays are added including SurfAbsSolarExt, SurfAbsThermalExt, SurfAbsSolarInt, SurfAbsThermalInt, SurfRoughnessExt. They are initialized with material properties one time at the beginning of InitSurfaceHeatBalance, and are set to movable insulation properties (and set back) at run time when AnyMovableInsulation`.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: `SurfMovInsulExtPresent`, `SurfMovInsulIntPresent`, `SurfMovInsulIntPresentPrevTS`, `SurfMovInsulHExt`, `SurfMovInsulHInt` are allocated only when `AnyMovableInsulation`, and whenever they are used (e.g. before checking if `SurfMovInsulExtPresent`), `AnyMovableInsulation` is checked first.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Deleted.

File: src/EnergyPlus/HeatBalanceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceManager.cc]: Flag added.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: A vector `SurfMovInsulIndexList` is initialized at the input stage and the loops are reduced.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Like explained in the previous comment, `AbsorpSolar` and `AbsorpThermal` are saved to runtime HB arrays. I removed `SurfMovInsulAbsInt` array here and added `SurfAbsSolarInt` instead.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: I added `SurfMovInsulExtPresent` and used the boolean arrays in checking presence for both inside and outside surface.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Used the index list instead.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Runtime `SurfAbsThermalExt` is used instead here and initialized with `state.dataSurface->SurfActiveConstruction` at the beginning of simulation or when any EMS override is on.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: In [develop branch](https://github.com/NREL/EnergyPlus/blob/b622e27ca3c96566d6bbc8c6429ae971f880d445/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L3570), `AbsIntSurfVis` is set as `Construct(ConstrNum).InsideAbsorpSolar` instead of `InsideAbsorpVis`, and is not reset when movable insulation is on. Although the comment here says "Inside opaque surface visible absorptance to fix CR 8695 change to this = Construct(ConstrNum)%InsideAbsorpVis". I wonder if this is a bug or it's intended to be fixed like this. @mjwitte Could you help take a look?

**Comment by xuanluo113** [comment]
@mjwitte Thanks for the review. I've finished addressing most of the comments as replied, except for two items to work on here or later.

1. The `AbsIntSurfVis` bug (?) as above.
2. `dataHeatBal->ITABSF` seems to have the same physical meaning as the newly added `SurfAbsThermalExt` array, while `ITABSF` is used in window surfaces and `SurfAbsThermalExt` in opaque surfaces. I intend to remove `ITABSF`, but this can be done along with the todo item listed above - reducing `ComputeIntThermalAbsorpFactors` and `ComputeIntSWAbsorpFactors`?

File: src/EnergyPlus/DataGlobals.hh
Comment (mjwitte) [inline src/EnergyPlus/DataGlobals.hh]: This endline comment should be more specific. And probably should be named `AnySurfPropOverridesInModel` to be consistent with the others? This can be be a TODO for later if you want.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Is it necessary to check for `ConstrNum>0`? All heat transfer surfaces should have a valid construction.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:3624]: If I'm following correctly, `AbsIntSurfVis` is only used for reporting, not in the actual surface heat gain calculations. Per the comments in [CR8695](https://github.com/NREL/EnergyPlus/issues/3468) it sounds like these needs to remain AbsSolar so that the reporting is consistent with the actual heat balance. But it's certainly worth looking at. It might be possible to fix that (later).

**Review by mjwitte — APPROVED** [review]
@xuanluo113 This is so much better with much cleaner logic. A few minor comments below.
I haven't looked at the ITABSF question yet.
I'll wait for CI to finish before merging. Or if you want to address the SurfPropOverridesPresent comment in this PR, let me know.

File: src/EnergyPlus/DataGlobals.hh
Comment (xuanluo113) [reply src/EnergyPlus/DataGlobals.hh]: I've fixed this comment and naming. Waiting for the local tests to finish for removing `ConstrNum > 0`. Will make a push later.

**Comment by mjwitte** [comment]
@xuanluo113 Looks good. Merging.

**Comment by jcyuan2020** [comment]
This PR gets an extra credit for fixing a newly reported issue #8758.

**Pull request merged**
